File: 141. Linked List Cycle/linked_list_cycle.py
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head == None:
            return False

        slow = head
        fast = head
        while fast.next != None and fast.next.next != None:
            slow = slow.next
            fast = fast.next.next

            if fast == slow:
                return True

        return False

# https://leetcode.com/problems/linked-list-cycle/

File: 141. Linked List Cycle/linked_list_cycle.dart
class ListNode {
  int val;
  ListNode? next;
  ListNode([this.val = 0, this.next]);
}

class Solution {
  bool hasCycle(ListNode? head) {
    if (head == null) return false;
    ListNode? slow = head;
    ListNode? fast = head;
    while (fast!.next != null && fast.next!.next != null) {
      slow = slow?.next;
      fast = fast.next!.next!;
      if (fast == slow) {
        return true;
      }
    }
    return false;
  }
}
//https://leetcode.com/problems/linked-list-cycle/